rfcs/0034-system-parameters.md

```diff
@@ -0,0 +1,135 @@
+
+### Session Parameters
+
+PostgreSQL allows users to set parameters for the current session as well.
+
+```sql
+SET parameter TO value / 'value' / DEFAULT;
+```
+
+References:
+
+- [PostgreSQL: Documentation: 15: 20.1. Setting Parameters](https://www.postgresql.org/docs/current/config-setting.html)
+
+Previously, we had already introduced [session parameters in RisingWave](https://github.com/risingwavelabs/risingwave/blob/53f7e0db772ac7e51773791bb8301624ed763ae8/src/common/src/session_config/mod.rs#L265). Some of them should be system parameters like `batch_enable_sort_agg` or `query_mode`. Postgres **enforces** that session parameters must also be system paramters, we can follow this rule as well.
+
+### Process Configs
+
+Let's rethink the current design of CLI options and config files. ([#5676](https://github.com/risingwavelabs/risingwave/issues/5676)) Given the existence of system parameters, some existing parameters may need to be moved to system parameters, like `data_directory` as mentioned before.
+
+We propose to consider CLI options as a way to override config files (i.e. in term of priority, CLI options -> config file -> defaults); as a result, the items in CLI options are a subset of the config file. The reasons for this include
+
+- All configs are still too many to be written as CLI options. (See [example.toml](https://github.com/risingwavelabs/risingwave/blob/main/src/config/example.toml)) 
+- While CLI configs are handy to set some frequently-used options like endpoint and memory.
+
+This pattern is widely adopted in many databases and even other applications. For example, TiDB considers configuration file as a superset of CLI options:
+
+> The TiDB configuration file supports more options than command-line parameters. You can download the default configuration file [config.toml.example](https://github.com/pingcap/tidb/blob/master/config/config.toml.example) and rename it to `config.toml`. This document describes only the options that are not involved in [command line options](https://docs.pingcap.com/tidb/stable/command-line-flags-for-tidb-configuration).
+>
+> Ref. [TiDB Configuration File | PingCAP Docs](https://docs.pingcap.com/tidb/stable/tidb-configuration-file)
+
+In RisingWave it will be like
+
+```toml
+# risingwave.toml
+[server]
+total_memory_bytes = 1073741824
+parallelism = 4
+[storage]
+shared_buffer_capacity_mb = 4096
+sstable_size_mb = 256
+block_size_kb = 1024
+```
+
+```bash
+# Override total_memory_bytes as 2GB
+./risingwave compute-node --total-memory-bytes=2147483648
+```
+
+### Mutability
+
+- Process configs must be **immutable**, because there is no user-interface to change it.
+- Global configs might be **immutable or mutable**, the mutable ones can be further divided into take effect immediately or need reboot.
+    - **Immutable** e.g. `data_directory` should be rejected upon SET
+    - **Mutable & need reboot** e.g. `heartbeat_interval_ms`
+    - **Mutable & immediately** e.g. `batch_chunk_size`
```

Comment:
I guess there're more things to consider in defining "immediately". As Flink says,

> Because options are read at different point in time when performing operations, it is recommended to set configuration options early after instantiating a table environment.

If we only read the configuration on scheduling (of batch queries...), then it sounds not that "immediate" 😄 but more like a session variable that only affects future queries. If we read it continuously, we may need to consider whether it'll lead to unexpected behavior if the config can be modified at any time.

Reply:
Exactly. The boundary seems obscure for these cases. I just collectively call them "immediately" to distinguish them from the need-reboot ones.